android-components/components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt

```diff
@@ -212,12 +212,21 @@ class DefaultAddonUpdater(
         logger.info("onUpdatePermissionRequest $current")
 
         val shouldGrantWithoutPrompt = Addon.localizePermissions(newPermissions, applicationContext).isEmpty()
-        val shouldShowNotification =
+        val shouldNotPrompt =
             updateStatusStorage.isPreviouslyAllowed(applicationContext, updated.id) || shouldGrantWithoutPrompt
 
-        onPermissionsGranted(shouldShowNotification)
-
-        if (!shouldShowNotification) {
+        // When the extension update doesn't have new permissions that the user should grant with a prompt,
+        // we allow the update to continue.
+        //
+        // Otherwise, the permission request will first be user-cancelled because we return `false` below
+        // but we create an Android notification right after, which is responsible for prompting the user.
+        // When the user allows the new permissions in the Android notification, the extension update is
+        // triggered again and - since the permissions have been previously allowed - there is no new
+        // permissions that the user should grant and so we allow the update to continue. At this point,
+        // the extension is fully updated.
+        onPermissionsGranted(shouldNotPrompt)
+
+        if (!shouldNotPrompt) {
```

Comment:
It's bit hard to read, as now we are add a negation `!` on variable that includes negation `Not` as part of its name.